sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -3305,7 +3321,24 @@ pi_result piEventsWait(pi_uint32 NumEvents, const pi_event *EventList) {
   for (uint32_t I = 0; I < NumEvents; I++) {
     ze_event_handle_t ZeEvent = EventList[I]->ZeEvent;
     zePrint("ZeEvent = %lx\n", pi_cast<std::uintptr_t>(ZeEvent));
-    ZE_CALL(zeEventHostSynchronize(ZeEvent, UINT32_MAX));
+
+    // If event comes from a Map/UnMap operation in integrated device, then do
+    // sync, memcpy, and signaling on the host
+    if (EventList[I]->HostSyncforMap) {
+      for (auto ZeWaitEvent : EventList[I]->waitEvents) {
+        zePrint("ZeWaitEvent = %lx\n", pi_cast<std::uintptr_t>(ZeWaitEvent));
+        if (ZeWaitEvent)
+          ZE_CALL(zeEventHostSynchronize(ZeWaitEvent, UINT32_MAX));
+      }
+      if (EventList[I]->CopyPending) {
+        memcpy(EventList[I]->DstBuffer, EventList[I]->SrcBuffer,
+               EventList[I]->RetMapSize);
+        EventList[I]->CopyPending = false;
+      }
```

Comment:
Why is copy done in event's wait not in map/unmap? Looks wrong.

Reply:
The intent here is to delay the copy until it is needed. The existing code was essentially doing that, by pushing a copy command onto a commandlist and returning the output event of that copy. The Map function returned immediately and the completion of the copy was enforced only when someone waited on the output event. The same thing is happening now.
I can try doing a synchronous copy in this function but that I suspect would affect performance.

Reply:
Please do the copy in map/unmap unless it really is a problem. It can actually be even faster if the copy is already completed before its results are really needed.

Reply:
In the map/unmap operations there is an incoming events list.
In the old code, within the map/unmap functions we append the following to a commandlist:
1. the incoming events of map/unmap
2. a copy operation needed for the map/unmap 
Then the commandlist is executed with the output event of the execution set as the output event of the map/unmap operation. This ensures that incoming events are satisfied, then the copy is done, and then the map/unmap operation is signaled as complete.
Solution1. In the implementation put up for review, 1. and 2. above are deferred until the map/unmap operation is waited upon. That's why we save all the information needed to do 1. and 2. in new fields in the event structure, and when the output event of map/unmap is waited upon, we do the actions in 1. and 2 above.
Solution2. You are asking that we do an immediate copy in the map/unmap function. To do that we must wait upon the incoming operations of the map/unmap function before doing a synchronous copy. There may be commandlists associated with those events, so those commandlists must be executed. This was already being done in the EventsWait function.
Solution 1 moved the copy functionality from map/unmap functions into EventsWait.
Solution 2 requires moving the execution of commandlists associated with incoming events of Map/Unmap into the Map/Unmap function. 
I have implemented both under an #if in pi_level_zero.hpp and in EventsWait, Map and Unmap functions in pi_level_zero.cpp. With #if 1 we get Solution 1. With #if 0 we get Solution 2. The performance of both is about the same.
Please review and tell me which you prefer.

Reply:
Thanks for reminding that we need to wait for input events before doing the copy. Please follow your Solution2, and just call piEventsWait for the input events.

Reply:
OK